src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -1973,45 +2940,109 @@ public UncaughtExceptionHandler getUncaughtExceptionHandler() {
+     * called when a thread terminates with an exception.
      */
-    private void dispatchUncaughtException(Throwable e) {
+    void dispatchUncaughtException(Throwable e) {
         getUncaughtExceptionHandler().uncaughtException(this, e);
     }
 
+    /**
+     * Holder class for constants.
+     */
+    @SuppressWarnings("removal")
+    private static class Constants {
+        // Thread group for virtual threads.
+        static final ThreadGroup VTHREAD_GROUP;
+
+        // AccessControlContext that doesn't support any permissions.
+        @SuppressWarnings("removal")
+        static final AccessControlContext NO_PERMISSIONS_ACC;
+
+        // Placeholder TCCL when thread locals not supported
+        static final ClassLoader NOT_SUPPORTED_CLASSLOADER;
+
+        static {
+            var getThreadGroup  = new PrivilegedAction<ThreadGroup>() {
+                @Override
+                public ThreadGroup run() {
+                    ThreadGroup parent = Thread.currentCarrierThread().getThreadGroup();
+                    for (ThreadGroup p; (p = parent.getParent()) != null; )
+                        parent = p;
+                    return parent;
+                }
+            };
+            @SuppressWarnings("removal")
+            ThreadGroup root = AccessController.doPrivileged(getThreadGroup);
+            VTHREAD_GROUP = new ThreadGroup(root, "VirtualThreads", NORM_PRIORITY, false);
+
+            NO_PERMISSIONS_ACC = new AccessControlContext(new ProtectionDomain[] {
+                new ProtectionDomain(null, null)
+            });
+
+            var createClassLoader = new PrivilegedAction<ClassLoader>() {
+                @Override
+                public ClassLoader run() {
+                    return new ClassLoader(null) { };
+                }
+            };
+            @SuppressWarnings("removal")
+            ClassLoader loader = AccessController.doPrivileged(createClassLoader);
+            NOT_SUPPORTED_CLASSLOADER = loader;
+        }
+    }
+
     // The following three initially uninitialized fields are exclusively
     // managed by class java.util.concurrent.ThreadLocalRandom. These
     // fields are used to build the high-performance PRNGs in the
-    // concurrent code, and we can not risk accidental false sharing.
-    // Hence, the fields are isolated with @Contended.
+    // concurrent code.
 
     /** The current seed for a ThreadLocalRandom */
-    @jdk.internal.vm.annotation.Contended("tlr")
     long threadLocalRandomSeed;
 
     /** Probe hash value; nonzero if threadLocalRandomSeed initialized */
-    @jdk.internal.vm.annotation.Contended("tlr")
     int threadLocalRandomProbe;
 
     /** Secondary seed isolated from public ThreadLocalRandom sequence */
-    @jdk.internal.vm.annotation.Contended("tlr")
     int threadLocalRandomSecondarySeed;
 
+    /** The thread container that this thread is in */
+    @Stable private volatile ThreadContainer container;
```

Comment:
The `volatile` modifier with `@Stable` is unusual. IIUC the field is set once on start (behaves as if a final field). Once C2 gets hold of it and can treat the value as a runtime constant the `volatile` has no effect. This is a tricky area but i think a store fence will suffice in `setThreadContainer` since the value goes from null to non-null and therefore cannot be speculated upon.